extensions/azure/processors/PutAzureBlobStorage.cpp

```diff
@@ -138,38 +149,75 @@ void PutAzureBlobStorage::onSchedule(const std::shared_ptr<core::ProcessContext>
-  return credentials.getConnectionString();
+  std::string value;
+  if (context->getProperty(StorageAccountName, value, flow_file)) {
+    credentials.setStorageAccountName(value);
+  }
+  if (context->getProperty(StorageAccountKey, value, flow_file)) {
+    credentials.setStorageAccountKey(value);
+  }
+  if (context->getProperty(SASToken, value, flow_file)) {
+    credentials.setSasToken(value);
+  }
+  if (context->getProperty(CommonStorageAccountEndpointSuffix, value, flow_file)) {
+    credentials.setEndpontSuffix(value);
+  }
+  if (context->getProperty(ConnectionString, value, flow_file)) {
+    credentials.setConnectionString(value);
+  }
+  credentials.setUseManagedIdentityCredentials(use_managed_identity_credentials_);
+  return credentials;
 }
 
-void PutAzureBlobStorage::createAzureStorageClient(const std::string &connection_string, const std::string &container_name) {
-  // When used in multithreaded environment make sure to use the azure_storage_mutex_ to lock the wrapper so the
-  // client is not reset with different configuration while another thread is using it.
-  if (blob_storage_wrapper_ == nullptr) {
-    blob_storage_wrapper_ = std::make_unique<storage::AzureBlobStorage>(connection_string, container_name);
-    return;
+std::optional<storage::PutAzureBlobStorageParameters> PutAzureBlobStorage::buildAzureBlobStorageParameters(
+    const std::shared_ptr<core::ProcessContext> &context,
+    const std::shared_ptr<core::FlowFile> &flow_file) {
+  storage::PutAzureBlobStorageParameters params;
+  auto credentials = getCredentials(context, flow_file);
+  if (!credentials) {
+    logger_->log_error("No valid credentials are set!");
+    return std::nullopt;
+  }
+
+  params.credentials = *credentials;
+
+  if (!context->getProperty(ContainerName, params.container_name, flow_file) || params.container_name.empty()) {
+    logger_->log_error("Container Name is invalid or empty!");
+    return std::nullopt;
+  }
+
+  context->getProperty(Blob, params.blob_name, flow_file);
+  if (params.blob_name.empty() && (!flow_file->getAttribute("filename", params.blob_name) || params.blob_name.empty())) {
+    logger_->log_error("Blob is not set and default 'filename' attribute could not be found!");
+    return std::nullopt;
   }
 
-  blob_storage_wrapper_->resetClientIfNeeded(connection_string, container_name);
+  return params;
 }
 
-std::string PutAzureBlobStorage::getConnectionString(
+std::optional<storage::AzureStorageCredentials> PutAzureBlobStorage::getCredentials(
     const std::shared_ptr<core::ProcessContext> &context,
     const std::shared_ptr<core::FlowFile> &flow_file) const {
-  auto connection_string = getAzureConnectionStringFromProperties(context, flow_file);
-  if (!connection_string.empty()) {
-    return connection_string;
+  auto controller_service_creds = getCredentialsFromControllerService(context);
+  if (controller_service_creds && controller_service_creds->isValid()) {
```

Comment:
I feel like we should differentiate between "nullopt because the user didn't set the Azure Storage Credentials Service property" and between the "user set something (explicitly requesting the service) but did a typo or added a different service by accident", in the latter case I don't think that the fallback to property credentials is the way to go

Reply:
a default constructed AzureStorageCredentials seems to be inValid, that could be returned in the second case

Reply:
Good point, added a change to fail in case invalid parameters are set in the credentials service in c9965a1249b9fae617c6a4467d1ec2596b915158

Reply:
we might need to touch `getCredentialsFromControllerService` as well, to return not a nullopt but an invalid `AzureStorageCredentials` after it observes that the `Azure Storage Credentials Service` is set but subsequent attempts to get the service fail

Reply:
I'm not sure if it's worth differentiating if the provided service name is invalid or empty. From the user's point of view the result will be the same and you will get a specific error message which was the case here. Do you think it's a problem that in both cases we check if the property credentials are set?

Reply:
right now it seems to me, that if the user set the `Azure Storage Credentials Service` property but there is no such service or it is not of the appropriate type we return `std::nullopt`, and we go on trying to create the credentials from the properties despite the user explicitly providing the service name

Reply:
Yes I was just conflicted if that was actually a bad thing. But I suppose technically it could happen that the service name is wrong and the property credentials are set to some valid parameters and the processor does not use the credentials it was intended to use and works regardless. I added a result code to the function in 2c5a10d585fabbd6f2127ad9a0814dcb4bcee719. I think that more explicitly shows the cause of failure.